utilities/include/vec4.hpp

```diff
@@ -0,0 +1,149 @@
+#ifndef VEC4_HPP
+#define VEC4_HPP
+
+#include <ostream>
+#include <cmath>
+#include <xsimd/xsimd.hpp>
+
+#define ALPHA_ 0 //include alpha or not.
+
+#if XSIMD_WITH_AVX512BW || XSIMD_WITH_AVX512DQ || XSIMD_WITH_AVX512F
```

Comment:
AVX512 has a different register length than AVX. If the user insists on using 512, that sets the alignment to 64. Also the macros XSIMD are set to 1 whenever they are available at compile time. We still don't have a robust mechanism to auto magically deduce the ISA to use or let the user specify the ISA.
Also I don't understand why do we need to make those "f" classes explicitly

Reply:
We are the end users of the `vec4` class. Whatever ISA we say the processor has to use, it must use that set itself. Inside the ray tracer code, the programmer will be more occupied with figuring out the ray tracing stuff, rather than what ISA the `vec4` class should be using.

Also, using AVX 512 for something like this would actually lead to a massive performance drop because most of the vector lanes will be unused.

The reason I was saying that you might want to make a `vec4f` class was because `float`s are 32 bits, so 4 floats can fit perfectly in a 128-bit register. However, `double`s are 64 bits so 4 doubles can only fit in a 256 bit register. The former is supported by SSE, whereas the latter requires AVX minimum. For this reason, having different names for the classes will help (because, since they are different classes, you can switch the implementation based on the floating point data type size.)

Reply:
is there a way to switch between two classes based on ISA? We can again create a conditional macro, but is there a better way? I think we can throw an error if the user tries to use Doubles on a machine that doesn't support AVX rather than repeating code in a `vec4f`

Reply:
I believe the performance would suffer since a large part of cache would be garbage?

Reply:
> is there a way to switch between two classes based on ISA? We can again create a conditional macro, but is there a better way? I think we can throw an error if the user tries to use Doubles on a machine that doesn't support AVX rather than repeating code in a `vec4f`

We can look into doing compile-time dispatch using `std::enable_if`, but I think we shouldn't complicate the implementation too much now itself.

Reply:
I have created a branch and pushed it in my fork [config-branchinator](https://github.com/prateekgargX/Rendera/tree/config-branchinator). Check it out. I'll be incorporating other changes in this branch. 
Also, we would need to place those files in the correct directory structure as well.